server.py

```diff
@@ -0,0 +1,149 @@
+    """
+    @wraps(f)
+    def decorated(*args, **kwargs):
+        try:
+            token = get_token_auth_header()
+            jsonurl = urlopen("https://login.microsoftonline.com/" +
+                            TENANT_ID + "/discovery/v2.0/keys")
+            jwks = json.loads(jsonurl.read())
+            unverified_header = jwt.get_unverified_header(token)
+            rsa_key = {}
+            for key in jwks["keys"]:
+                if key["kid"] == unverified_header["kid"]:
+                    rsa_key = {
+                        "kty": key["kty"],
+                        "kid": key["kid"],
+                        "use": key["use"],
+                        "n": key["n"],
+                        "e": key["e"]
+                    }
+        except Exception:
+            raise AuthError({"code": "invalid_header",
+                                "description":
+                                "Unable to parse authentication"
+                                " token."}, 401)
+        if rsa_key:
+            try:
+                payload = jwt.decode(
+                    token,
+                    rsa_key,
+                    algorithms=["RS256"],
+                    audience=API_AUDIENCE,
+                    issuer="https://sts.windows.net/" + TENANT_ID + "/"
+                )
+            except jwt.ExpiredSignatureError:
+                raise AuthError({"code": "token_expired",
+                                 "description": "token is expired"}, 401)
+            except jwt.JWTClaimsError:
+                raise AuthError({"code": "invalid_claims",
+                                 "description":
+                                 "incorrect claims,"
+                                 "please check the audience and issuer"}, 401)
+            except Exception:
+                raise AuthError({"code": "invalid_header",
+                                 "description":
+                                 "Unable to parse authentication"
+                                 " token."}, 401)
+            _request_ctx_stack.top.current_user = payload
+            # print(_request_ctx_stack.top.current_user)
+            return f(*args, **kwargs)
+        raise AuthError({"code": "invalid_header",
+                         "description": "Unable to find appropriate key"}, 401)
+    return decorated
+
+
+def requires_scope(required_scope):
+    """Determines if the required scope is present in the Access Token
+    Args:
+        required_scope (str): The scope required to access the resource
+    """
+    token = get_token_auth_header()
+    unverified_claims = jwt.get_unverified_claims(token)
+    if unverified_claims.get("scope"):
+            token_scopes = unverified_claims["scope"].split()
+            for token_scope in token_scopes:
+                if token_scope == required_scope:
+                    return True
+    return False
+
+# Controllers API
+
+# This doesn't need authentication
+@app.route("/api/public")
```

Comment:
Can this be decoupled into unit tests. This route is only for the purpose of demo and can be removed from the main code,